Improve signature help by more stable position calculation + better named arg support

rochala opened the pull request:
Previously, signature help had unstable way of calculating active parameter inside the signature help. It is now changed to work better with erroneous trees such as unclosed openings.

It also adds a new feature for signature help, which will help user navigate inside named arguments. Furthermore, it will now find reordered named arguments, and display in the same way, by also marking the remaining parameters that they are now required to be named.

Example:

https://github.com/lampepfl/dotty/assets/48657087/b181d2d5-60f0-46a5-b2df-a58aa5f07454

For the reviewers: I recommend testing it locally.

kasiaMarek commented on presentation-compiler/src/main/dotty/tools/pc/SignatureHelpProvider.scala:
````diff
@@ -1,79 +1,134 @@
+   *  // eof or illegal token
+   *  ```
+   *  we need to rely on dianostics to locate last tree before the cursor
+   *  and `opnning closed` expected syntax error.
````
```suggestion
   *  and `opening closed` expected syntax error.
```

kasiaMarek commented on presentation-compiler/src/main/dotty/tools/pc/SignatureHelpProvider.scala:
````diff
@@ -1,79 +1,134 @@
+   *  @@
+   *  // eof or illegal token
+   *  ```
+   *  we need to rely on dianostics to locate last tree before the cursor
````
```suggestion
   *  we need to rely on diagnostics to locate last tree before the cursor
```

kasiaMarek commented on presentation-compiler/src/main/dotty/tools/pc/SignatureHelpProvider.scala:
```diff
@@ -1,79 +1,134 @@
+   *  When there is an unclosed opening, parser will try to repair it, but the span of that tree
+   *  will end on the new line or next illegal token.
+   *
+   *  To support getting the tree withh unclosed opening, when our cursor is on the newline such as:
```
```suggestion
   *  To support getting the tree with unclosed opening, when our cursor is on the newline such as:
```

kasiaMarek reviewed (commented):
Also the unclosed opening works for params but not type params.

jkciesluk commented on presentation-compiler/src/main/dotty/tools/pc/SignatureHelpProvider.scala:
```diff
@@ -1,79 +1,134 @@
-    val trees = driver.openedTrees(uri.nn)
-
-    val path = Interactive.pathTo(trees, pos)
+    driver.compilationUnits.get(uri) match
```
Why this change?

rochala replied:
We need untpd trees, and without this you don't have this information in context.

jkciesluk reviewed (changes requested):
Great job! It fixes a lot of things, but I'm unsure if handling unclosed openings is needed, especially because it really complicates finding enclosing apply.

Could you add these tests:
```scala
def test[T, F](aaa: Int, bbb: T, ccc: F): T = ???
val x = test(1, ccc = 2, b@@)
```

```scala
def test2(aaa: Int, bbb: Int, ccc: Int): Int = ???
val x = test2(
    1, 
    @@
  )
```

```scala
def test2(aaa: Int, bbb: Int, ccc: Int): Int = ???
val x = test2(aaa = 2@@, ccc = 1)
```

```scala
def test2(aaa: Int, bbb: Int, ccc: Int): Int = ???
val x = test2(aaa = 2, @@  ccc = 1, bbb = 3)
```

Also, would it be possible to show in the case below that arguments in the first parameter list are reordered?

```scala
def test3(aaa: Int, bbb: Int, ccc: Int)(ddd: Int): Int = ???
val x = test3(bbb = 1, aaa = 2, ccc = 3)(d@@)
```

rochala commented:
The following changes have been made:
- we will stop supporting Signature Helps for unclosed openings. It is too hard to recover from such errors reliably and to track where we are with the cursor. All modern editors will close opening on input. It simplified the code, and is the best way forward.
- if there is an error within a definition return type parameter, we will now fall back to source, instead of print error,
- tuples and functions are now supported when they are returned from methods.
- this rewrite made me change a few things, and I took the chance and made it support clause interleaving. There is now a proper test suite for that. It required changing the way we represent Signatures.
- argument reordering now works for all parameter lists, not only the currently applied one,
- type parameters now also correctly display documentation

rochala commented:
I've changed the PR to use `ShortenedTypePrinter` and fixed some other minor issues like correctly marking reordered argumentss or hiding synthetic parameter names. 

In the future, Signature help should be changed to return symbols from the compiler instead of already printed values, and only print them later using shared API of ShortenedTypePrinter in the Presentation Compiler. Ideally we'd want a shared implementation for label printing with completions, hovers and other features.

jkciesluk reviewed (approved):
LGTM! Great work!

rochala commented on compiler/src/dotty/tools/dotc/util/Signatures.scala:
```diff
@@ -397,85 +452,136 @@ object Signatures {
+
+      def reduceToParamss(applies: List[untpd.Tree], types: List[Type], paramList: Int = 0): List[List[Param]] =
+        applies -> types match
+          case ((_: untpd.TypeApply) :: restTrees, (poly: PolyType) :: restTypes) =>
```
It matches against non empty apply list. Removing it failed the test.

rochala commented on presentation-compiler/src/main/dotty/tools/pc/SignatureHelpProvider.scala:
```diff
@@ -1,134 +1,149 @@
+
+        given Context = localizedContext.fresh
+          .setCompilationUnit(unit)
+          .setPrinterFn(_ => ShortenedTypePrinter(search, IncludeDefaultParam.Include)(using indexedContext))
```
This should be not include

kasiaMarek commented on presentation-compiler/test/dotty/tools/pc/tests/signaturehelp/SignatureHelpNamedArgsSuite.scala:
```diff
@@ -0,0 +1,300 @@
+         |""".stripMargin
+    )
+
+  @Test def `new-named-param-style-double-space` =
```
Unclosed opening test?

kasiaMarek commented on language-server/src/dotty/tools/languageserver/DottyLanguageServer.scala:
```diff
@@ -930,23 +935,25 @@ object DottyLanguageServer {
-          }
-          .mkString("(", ")(", ")")
-    val tparamsLabel = if (signature.tparams.isEmpty) "" else signature.tparams.mkString("[", ", ", "]")
+      signature.paramss
```
I assume this whole duplication is because there is no good place for it, such that it could be used by both modules.

rochala replied:
Dotty language server should actually be removed from the repository as it is no longer necessary. It is not so easy tho as the test coverage is different. This change was only made to make the tests pass

kasiaMarek commented on presentation-compiler/src/main/dotty/tools/pc/SignatureHelpProvider.scala:
```diff
@@ -1,134 +1,149 @@
+    driver.compilationUnits.get(uri) match
+      case Some(unit) =>
+
+        val pos = driver.sourcePosition(params, isZeroExtent = false)
```
If you're expanding offset to a range span, won't that throw an error later if the cursor is at the very end of the file? I know, we don't do open brackets but we should make sure not to throw.

kasiaMarek reviewed (approved):
Looks really good. 🚀  I feel like you improved naming of different variables/methods, thanks for that, I think it made things more readable.

rochala commented on presentation-compiler/test/dotty/tools/pc/tests/signaturehelp/SignatureHelpPatternSuite.scala:
```diff
@@ -202,23 +202,6 @@ class SignatureHelpPatternSuite extends BaseSignatureHelpSuite:
          |""".stripMargin
     )
 
-  @Test def `pat4` =
```
A. It is not working any more due to changes to range.
B. It was a bit useless as it would not be automatically opened and triggered, also the information is not that helpful, especially if you have more than single Apply like
`a :: b :: c :: Nil`

rochala commented on presentation-compiler/test/dotty/tools/pc/tests/signaturehelp/SignatureHelpSuite.scala:
```diff
@@ -323,9 +323,9 @@ class SignatureHelpSuite extends BaseSignatureHelpSuite:
         |}
       """.stripMargin,
-      """|to(end: Int): scala.collection.immutable.Range.Inclusive
+      """|to(end: Int): Inclusive
```
No, it should only shorten packages. I'll change it in separate PR as this is not caused by these changes.
